path_calc.hpp

```diff
@@ -0,0 +1,83 @@
+#ifndef IRODS_INDEXING_PATH_CALC_HPP
+#define IRODS_INDEXING_PATH_CALC_HPP
+
+#include <set>
+#include <map>
+#include <string>
+#include <functional>
+
+using namespace std::string_literals;
+
+namespace irods {
+    namespace indexing {
+
+        // Thrown by class path_calc_and_cache, when given a path
+        //  that is not well-formed. This will result from any attempt
+	//  to ascend a collection hierarchy not starting with "/".
+
+        class  path_format_error: public std::runtime_error {
+        public:
+            path_format_error(const std::string & s) : std::runtime_error(s.c_str()) {}
+        };
+
+        // Calculates the cumulative set of items of the given type Property
+        // that are applicable at a given level of the iRODS collection hierarchy.
+
+        template <typename Property>
+        class path_calc_and_cache {
+
+        public:
+            using properties = std::set<Property>;
+
+        private:
+            std::map<std::string, properties> path_properties;
+            using calc_function = std::function<properties (const std::string& s)> ;
+            calc_function calc_;
+
+        public:
+
+            // Initialize with a callable object that returns the Property value for a given collection path
+            // within the hierarchy.
+            //
+            explicit path_calc_and_cache(calc_function f)
+                : calc_{f}
+            {
+            }
+
+            properties accum(const std::string &path)
+            {
+                auto pos = (path.size() < 1 ? std::string::npos : path.find_last_of("/"));
+                if (pos == std::string::npos) {
+                    throw path_format_error{ "Couldn't parse: "s + path };
+                }
+                if (path == "/") {
+                    return calc("/");
+                }
+                auto parent_path_length = (pos == 0) ? 1 : pos;
+                auto accumulated_set = accum(path.substr(0,parent_path_length));
+                auto current_path_set = calc(path);
+                accumulated_set.insert (current_path_set.begin(), current_path_set.end());
+                return accumulated_set;
+            }
+
+            // Get individual contribution for last elem of path 'p'.
+            // Use cached value if it exists, else do the calculation
+            properties calc(const std::string & path) {
+                properties partial;
+                try{
+                    partial = path_properties.at(path);
+                }
+                catch(const std::out_of_range &e) {
+                    partial = (this->calc_)(path);
+                }
+                path_properties[path] = partial;
+                return partial;
+            }
+
+        }; // class
```

Comment:
Missing class name?

Just pointing it out. I wasn't sure if this was on purpose or not.